brion/plugin/compartmentReportBinary.cpp

```diff
@@ -251,16 +348,101 @@ floatsPtr CompartmentReportBinary::loadFrame(const float timestamp) const
             const uint64_t targetOffset = offsets[i][j];
 
             for (uint16_t k = 0; k < numCompartments; ++k)
-                (*buffer)[k + targetOffset] = source[k + sourceOffset];
+                buffer[k + targetOffset] = source[k + sourceOffset];
         }
     }
 
     if (_header.byteswap)
     {
 #pragma omp parallel for
         for (ssize_t i = 0; i < ssize_t(_subNumCompartments); ++i)
-            lunchbox::byteswap((*buffer)[i]);
+            lunchbox::byteswap(buffer[i]);
+    }
+    return true;
+}
+
+bool CompartmentReportBinary::_loadFrameAIO(const float timestamp,
+                                            float* buffer) const
+{
+    if (_offsets[_subtarget].empty())
+        return false;
+
+    const size_t frameNumber = _getFrameNumber(timestamp);
+    const size_t frameOffset =
+        _header.dataBlockOffset +
+        _header.numCompartments * sizeof(float) * frameNumber;
+
+    if (!_subtarget)
+    {
+        // clang-format off
+        auto cb = readAsync(
+            {
+                _fileDescriptor,
+                buffer,
+                _header.numCompartments * sizeof(float),
+                frameOffset
+            }
+        );
+        // clang-format on
+
+        waitFor(&cb);
+        checkBytesRead(&cb);
+
+        if (_header.byteswap)
+        {
+#pragma omp parallel for
+            for (int32_t i = 0; i < _header.numCompartments; ++i)
+                lunchbox::byteswap(buffer[i]);
+        }
+        return true;
+    }
+
+    if (_subNumCompartments == 0)
+        return false;
+
+    const SectionOffsets& offsets = getOffsets();
+    const CompartmentCounts& compCounts = getCompartmentCounts();
+    std::vector<AIOReadData> readData;
+
+    for (size_t i = 0; i < _gids.size(); ++i)
+    {
+        for (size_t j = 0; j < offsets[i].size(); ++j)
+        {
+            const uint16_t numCompartments = compCounts[i][j];
+            const uint64_t sourceOffset = _conversionOffsets[i][j];
+            const uint64_t targetOffset = offsets[i][j];
+
+            // clang-format off
+            readData.push_back(                      
+                {
+                   _fileDescriptor,
+                    buffer + targetOffset,   // destination buffer
+                    numCompartments * sizeof(float), // size
+                    frameOffset + sourceOffset * sizeof(float) //  global offset
+                }
+            );
+            // clang-format on
+        }
     }
+
```

Comment:
Try to refactor the code to not repeat the lines 428-438. If you use the same aio code for both single and multiple operations it would be simple, for example you can use a different helper function to create the std::vector<AIOReadData> depending on the case. If the vector is empty you can return false (that would be the case _subNumCompatments == 0)